### PR TITLE
MODINREACH-84 Add proxy controller to get data directly from Inn Reach

### PR DESCRIPTION
**Pull request by andrew-khodyka**
## Purpose
Add proxy controller to get data directly from Inn Reach

## Approach

- Add controller to handle all requests with /d2r/** URI
- Add Inn Reach client to call Inn Reach API

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/external/client/feign/error/FeignErrorDecoder.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  @Override
+  public Exception decode(String s, Response response) {
+    if (HttpStatus.valueOf(response.status()).equals(HttpStatus.UNAUTHORIZED)) {
+      log.debug("Can't get InnReach access token. CentralServer authentication failed with status: {}", response.status());
+      return new BadCredentialsException("Can't get InnReach access token. Key/Secret pair is not valid");
+    }
+
+    log.debug("Inn Reach call failed with status: {}", response.status());
+    return new InnReachException("Inn Reach call failed failed with status: {}");
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/external/client/feign/error/FeignErrorDecoder.java]:
```suggestion
    return new InnReachException("Inn Reach call failed with status: " + response.status());
```

File: src/main/java/org/folio/innreach/external/service/impl/InnReachAuthServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+@Log4j2
+@Service
+public class InnReachAuthServiceImpl implements InnReachAuthService {
+
+  private final InnReachAuthClient innReachAuthClient;
+  private final Cache<String, AccessTokenDTO> accessTokenCache;
+
+  @Override
+  public AccessTokenDTO getAccessToken(AccessTokenRequestDTO tokenRequestDTO) {
+    var cachedAccessToken = accessTokenCache.getIfPresent(tokenRequestDTO.getCentralServerUri());
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/external/service/impl/InnReachAuthServiceImpl.java:26]: Would it be better to use Spring abstractions for caching like  @Cacheable with Caffeine as a provider?
Relevant documentation:
* https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.caching.provider
* https://docs.spring.io/spring-framework/docs/current/reference/html/integration.html#cache

Later on we can apply the same technic to other external calls like  querying mod-users

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -31,7 +31,7 @@ spring:
 folio:
   tenant:
     validation:
-      enabled: true
+      enabled: false
```
Comment (dmtkachenko) [inline src/main/resources/application.yml]: Could you explain why the validation is disabled? I don't think it's correct

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -8,7 +8,7 @@ spring:
     continue-on-error: true
     username: ${DB_USERNAME:postgres}
     password: ${DB_PASSWORD:postgres}
-    url: jdbc:postgresql://${DB_HOST:localhost}:${DB_PORT:5432}/${DB_DATABASE:db}
+    url: jdbc:postgresql://${DB_HOST:localhost}:${DB_PORT:5432}/${DB_DATABASE:postgres}
```
Comment (dmtkachenko) [inline src/main/resources/application.yml]: Was it a necessary change?

File: src/main/resources/application.yml (hunk 3 above)
Comment (andrew-khodyka) [reply src/main/resources/application.yml]: Sorry, this is my fault. Changed for local testing, forgot to change back

File: src/main/resources/application.yml (hunk 4 above)
Comment (andrew-khodyka) [reply src/main/resources/application.yml]: Sorry, this is my fault. Changed for local testing, forgot to change back

File: src/main/java/org/folio/innreach/external/service/impl/InnReachAuthServiceImpl.java (hunk 2 above)
Comment (andrew-khodyka) [reply src/main/java/org/folio/innreach/external/service/impl/InnReachAuthServiceImpl.java:26]: Yes you are right. I was thinking about Spring Cache but decided to use Guava as a temporary solution and discuss Spring cache with you later.

File: src/main/java/org/folio/innreach/external/client/feign/error/FeignErrorDecoder.java (hunk 1 above)
Comment (andrew-khodyka) [reply src/main/java/org/folio/innreach/external/client/feign/error/FeignErrorDecoder.java]: Thanks, will fix this

**Review by dmtkachenko — APPROVED** [review]

**Review by dmtkachenko — COMMENTED** [review]
New API URI has to be described in ModuleDescriptor-template.json otherwise it won't be accessible from UI
Not sure if it's possible to defined URI with a mask in the descriptor. Please consult RMB documentation regarding this point:
* https://github.com/folio-org/raml-module-builder/blob/master/README.md

If it's not possible then each endpoint has to described specifically like we did till this moment

**Pull request merged**